File: .github/scripts/llm.py
#!/usr/bin/env python3
"""统一 LLM 客户端，支持 Gemini / OpenAI / DeepSeek / Qwen"""

import os
import requests
from abc import ABC, abstractmethod


class BaseLLM(ABC):
    @abstractmethod
    def generate(self, prompt: str, temperature: float = 0.7) -> str:
        pass


class GeminiLLM(BaseLLM):
    def __init__(self, api_key: str, api_base: str = "https://generativelanguage.googleapis.com/v1beta", model: str = "gemini-2.0-flash"):
        self.api_key = api_key
        self.api_base = api_base.rstrip("/")
        self.model = model

    def generate(self, prompt: str, temperature: float = 0.7) -> str:
        url = f"{self.api_base}/models/{self.model}:generateContent?key={self.api_key}"
        safe_url = url.replace(self.api_key, "***") if self.api_key else url
        print(f"Gemini请求: {safe_url}")
        payload = {
            "contents": [{"parts": [{"text": prompt}]}],
            "generationConfig": {"temperature": temperature, "maxOutputTokens": 8192},
        }
        try:
            resp = requests.post(url, json=payload, timeout=60)
            resp.raise_for_status()
            data = resp.json()
            parts = data.get("candidates", [{}])[0].get("content", {}).get("parts", [])
            return parts[0].get("text", "") if parts else ""
        except Exception as e:
            safe_err = str(e).replace(self.api_key, "***") if self.api_key else str(e)
            print(f"Gemini API 失败: {safe_err}")
            return ""


class OpenAILLM(BaseLLM):
    def __init__(self, api_key: str, api_base: str = "https://api.openai.com/v1", model: str = "gpt-4o"):
        self.api_key = api_key
        self.api_base = api_base.rstrip("/")
        self.model = model

    def generate(self, prompt: str, temperature: float = 0.7) -> str:
        url = f"{self.api_base}/chat/completions"
        headers = {"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"}
        messages = [{"role": "user", "content": prompt}]
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": 4096,
        }
        try:
            resp = requests.post(url, json=payload, headers=headers, timeout=60)
            resp.raise_for_status()
            data = resp.json()
            return data["choices"][0]["message"]["content"]
        except Exception as e:
            safe_err = str(e).replace(self.api_key, "***") if self.api_key else str(e)
            print(f"OpenAI API 失败: {safe_err}")
            return ""


class DeepSeekLLM(OpenAILLM):
    def __init__(self, api_key: str, api_base: str = "https://api.deepseek.com", model: str = "deepseek-chat"):
        super().__init__(api_key, api_base, model)


class QwenLLM(OpenAILLM):
    def __init__(self, api_key: str, api_base: str = "https://dashscope.aliyuncs.com/compatible-mode/v1", model: str = "qwen-plus"):
        super().__init__(api_key, api_base, model)


PROVIDERS = {
    "gemini": GeminiLLM,
    "openai": OpenAILLM,
    "deepseek": DeepSeekLLM,
    "qwen": QwenLLM,
}


def create_llm() -> BaseLLM:
    """根据环境变量创建 LLM 客户端"""
    provider = os.environ.get("LLM_PROVIDER", "").lower()
    if not provider:
        raise ValueError("未配置 LLM_PROVIDER")

    api_key = os.environ.get("LLM_API_KEY", "")
    api_base = os.environ.get("LLM_API_BASE", "")
    model = os.environ.get("LLM_MODEL", "")

    if not api_key:
        raise ValueError(f"未配置 LLM_API_KEY")

    cls = PROVIDERS[provider]
    kwargs = {"api_key": api_key}
    if api_base:
        kwargs["api_base"] = api_base
    if model:
        kwargs["model"] = model
    return cls(**kwargs)

File: worker/lib/llm.ts
import type { LLMClient, LLMConfig } from '../types.js'

class GeminiLLM implements LLMClient {
  private apiKey: string
  private apiBase: string
  private model: string

  constructor(apiKey: string, apiBase = 'https://generativelanguage.googleapis.com/v1beta', model = 'gemini-2.0-flash') {
    this.apiKey = apiKey
    this.apiBase = apiBase.replace(/\/+$/, '')
    this.model = model
  }

  async generate(prompt: string, temperature = 0.7): Promise<string> {
    const url = `${this.apiBase}/models/${this.model}:generateContent?key=${this.apiKey}`
    const payload = {
      contents: [{ parts: [{ text: prompt }] }],
      generationConfig: { temperature, maxOutputTokens: 8192 },
    }
    try {
      const resp = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(60000),
      })
      if (!resp.ok) {
        const text = await resp.text().catch(() => '')
        throw new Error(`HTTP ${String(resp.status)} ${text.slice(0, 200)}`)
      }
      const data: { candidates?: { content?: { parts?: { text?: string }[] } }[] } = await resp.json()
      const parts = data.candidates?.[0]?.content?.parts || []
      const text = parts[0]?.text || ''
      if (!text) throw new Error('未返回内容')
      return text
    } catch (e) {
      console.error('Gemini API 失败:', e)
      throw new Error(`Gemini API 失败: ${(e as Error).message}`)
    }
  }
}

class OpenAILLM implements LLMClient {
  protected apiKey: string
  protected apiBase: string
  protected model: string

  constructor(apiKey: string, apiBase = 'https://api.openai.com/v1', model = 'gpt-4o') {
    this.apiKey = apiKey
    this.apiBase = apiBase.replace(/\/+$/, '')
    this.model = model
  }

  async generate(prompt: string, temperature = 0.7): Promise<string> {
    const url = `${this.apiBase}/chat/completions`
    const payload = {
      model: this.model,
      messages: [{ role: 'user' as const, content: prompt }],
      temperature,
      max_tokens: 4096,
    }
    try {
      const resp = await fetch(url, {
        method: 'POST',
        headers: { Authorization: `Bearer ${this.apiKey}`, 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(60000),
      })
      if (!resp.ok) {
        const text = await resp.text().catch(() => '')
        throw new Error(`HTTP ${String(resp.status)} ${text.slice(0, 200)}`)
      }
      const data: { choices?: { message?: { content?: string } }[] } = await resp.json()
      const text = data.choices?.[0]?.message?.content || ''
      if (!text) throw new Error('未返回内容')
      return text
    } catch (e) {
      console.error('OpenAI API 失败:', e)
      throw new Error(`OpenAI API 失败: ${(e as Error).message}`)
    }
  }
}

class DeepSeekLLM extends OpenAILLM {
  constructor(apiKey: string, apiBase = 'https://api.deepseek.com', model = 'deepseek-chat') {
    super(apiKey, apiBase, model)
  }
}

class QwenLLM extends OpenAILLM {
  constructor(apiKey: string, apiBase = 'https://dashscope.aliyuncs.com/compatible-mode/v1', model = 'qwen-plus') {
    super(apiKey, apiBase, model)
  }
}

const PROVIDERS = new Map<string, new (apiKey: string, apiBase?: string, model?: string) => LLMClient>([
  ['gemini', GeminiLLM],
  ['openai', OpenAILLM],
  ['deepseek', DeepSeekLLM],
  ['qwen', QwenLLM],
])

export function createLLM({ provider, apiKey, apiBase, model }: LLMConfig): LLMClient {
  const p = (provider || '').toLowerCase()
  if (!p) throw new Error('未配置 LLM_PROVIDER')
  if (!apiKey) throw new Error('未配置 LLM_API_KEY')
  const cls = PROVIDERS.get(p)
  if (!cls) throw new Error(`不支持的 LLM 提供商: ${provider}`)
  return new cls(apiKey, apiBase, model)
}
